src/emulator/storage/rules/manager.ts

```diff
@@ -1,52 +1,72 @@
 import * as chokidar from "chokidar";
-import * as fs from "fs";
 import { EmulatorLogger } from "../../emulatorLogger";
 import { Emulators } from "../../types";
-import { FirebaseError } from "../../../error";
 import { SourceFile } from "./types";
 import { StorageRulesIssues, StorageRulesRuntime, StorageRulesetInstance } from "./runtime";
+import { RulesConfig } from "..";
 
 /**
- * Loads and maintains a {@link StorageRulesetInstance} for a given source file. Listens for
- * changes to the file and updates the ruleset accordingly.
+ * Keeps track of the rules source file and maintains a generated ruleset for one or more storage
+ * resources.
+ * */
+export interface StorageRulesManager {
+  /** Sets source file for each resource using the rules previously passed in the constructor. */
+  start: () => Promise<StorageRulesIssues>;
+
+  /** Retrieves the generated ruleset for the resource. */
+  getRuleset: (resource: string) => StorageRulesetInstance | undefined;
+
+  /**
+   * Updates the source file and, correspondingly, the file watcher and ruleset for the resource.
+   */
+  setSourceFile: (rules: SourceFile, resource: string) => Promise<StorageRulesIssues>;
+
+  /** Deletes source file, ruleset, and removes listeners from all files for all resources. */
+  close: () => Promise<void>;
+}
+
+/**
+ * Creates either a {@link StorageRulesManagerImplementation} to manage rules for a single resource
+ * or a {@link StorageRulesManagerRegistry} for multiple resources.
+ */
+export function createStorageRulesManager(
+  rules: SourceFile | RulesConfig[],
+  runtime: StorageRulesRuntime
+): StorageRulesManager {
+  return Array.isArray(rules)
+    ? new StorageRulesManagerRegistry(rules, runtime)
+    : new StorageRulesManagerImplementation(rules, runtime);
+}
+
+/**
+ * Maintains a {@link StorageRulesetInstance} for a given source file. Listens for changes to the
+ * file and updates the ruleset accordingly.
  */
-export class StorageRulesManager {
+class StorageRulesManagerImplementation implements StorageRulesManager {
   private _sourceFile?: SourceFile;
   private _ruleset?: StorageRulesetInstance;
   private _watcher = new chokidar.FSWatcher();
   private _logger = EmulatorLogger.forEmulator(Emulators.STORAGE);
 
-  constructor(private _runtime: StorageRulesRuntime) {}
+  constructor(private _initRules: SourceFile, private _runtime: StorageRulesRuntime) {}
+
+  async start(): Promise<StorageRulesIssues> {
+    return this.setSourceFile(this._initRules);
```

Comment:
How does a client use this class? Is it: 

```
const m = new StorageRulesManagerImplementation(initialRules);
m.start();
m.stop();
m.setSourceFile(newRules);
m.start();
m.stop();
```

Or is it:

```
const m = new StorageRulesManagerImplementation(initialRules);
m.start();
m.setSourceFile(newRules);
m.setSourceFile(otherRules);
m.stop();
```

Reply:
Another question is why do we need `start` at all? I understand the need for cleanup but why not have the manager start watching the file on init or when the first `SourceFile` is set?

Reply:
The latter, but nothing bad will happen if you do the former.

Reply:
Talked offline; will set `this._initRules` (rename to `this._rules`) in `setSourceFile` to make `start` idempotent